### ISSUE TITLE
kv: long running elastic workloads cause foreground workloads to stall

### ISSUE THREAD

**Issue by andrewbaptist** [cockroachdb/cockroach#134668]
**Describe the problem**

Running a long running elastic workload can cause a foreground workload to stall after a period of time. This occurs with RACv2 in `apply_to_elastic` mode. 

**To Reproduce**

To manually reproduce, start a fixed throughput workload with normal priority that uses ~50% of CPU. Observe stable throughput and latency. Then start an elastic workload for 30m with low priority. The elastic workload should be relatively low and fixed concurrency but enough that the cluster won't be able to process all of it. After ~15m the foreground workload will begin to be impacted.

To automatically reproduce run this command. In case the system change.

```
COCKROACH_RANDOM_SEED=5128980479178314991 roachtest run perturbation/metamorphic/elasticWorkload
```
This starts the test with the following variations:
```
test variations are: seed: 5128980479178314991, 
fillDuration: 10m0s, 
maxBlockBytes: 4096, 
perturbationDuration: 30m0s, 
validationDuration: 5m0s, 
ratioOfMax: 0.500000, 
splits: 10000, 
numNodes: 12, 
numWorkloadNodes: 1, 
vcpu: 8, 
disks: 1, 
memory: high, 
leaseType: epoch, 
cloud: gce, 
perturbation: {}
```

About 10-15 minutes into the perturbation the foreground workload becomes "unstable" and total throughput on the cluster drops well below the initial throughput without the elastic workload running. For periods of up to 5s there is no throughput on the cluster.

On at least one node we get high goroutine scheduling latency, and on three different nodes the WAL fsync latency jumps to 15s.

**Expected behavior**

AC should prevent the background workload from impacting the foreground workload. 

**Additional data / screenshots**

* [A recent run](https://github.com/cockroachdb/cockroach/issues/134584#issue-2642208726)
* [Grafana link](https://grafana.testeng.crdb.io/d/crdb-home/crdb-home?orgId=1&var-DS_PROMETHEUS=Prometheus&var-test_run_id=teamcity-17632465&var-test_name=perturbation-metamorphic-elasticworkload&from=1731010692290&to=1731014252516)

**Environment:**
 - CockroachDB version - 24.3 and master

**Comment by andrewbaptist** [comment]
An example trace from a slow operation. Note that the time is primarily spent in raft waiting for a local `MsgAppResp`. 

The test intervals are:
```
2024/11/08 15:41:38 admission_control_latency.go:925: Baseline interval     : 2024-11-08 15:04:02.656721444 +0000 UTC m=+1080.216014085 -> 2024-11-08 15:06:32.656721444 +0000 UTC m=+1230.216014085
2024/11/08 15:41:38 admission_control_latency.go:926: Perturbation interval : 2024-11-08 15:06:32.656920907 +0000 UTC m=+1230.216213547 -> 2024-11-08 15:36:38.918449531 +0000 UTC m=+3036.477742171
2024/11/08 15:41:38 admission_control_latency.go:927: Recovery interval     : 2024-11-08 15:36:38.919538717 +0000 UTC m=+3036.478831357 -> 2024-11-08 15:41:38.919538717 +0000 UTC m=+3336.478831357
```
And the grafana log can be seen here: 
https://grafana.testeng.crdb.io/d/J-yAVzkVddd/detailed?orgId=1&from=1731079800000&to=1731080400000&var-cluster=baptistcockroachlabscom-1731077162-01-n13cpu8hm&var-DS_PROMETHEUS=v9Zz2K6nz&var-node=All&var-rate_interval=30s&var-token_type=eval

[2024-11-08T15_34_55Z-UPSERT-2877.287ms.zip](https://github.com/user-attachments/files/17681010/2024-11-08T15_34_55Z-UPSERT-2877.287ms.zip)

**Comment by sumeerbhola** [comment]
Something odd is going on here.

The CPU is high, and there is some slot exhaustion, but it does not change around 20:50, when the instability starts: 
<img width="849" alt="Screenshot 2024-11-08 at 2 32 04 PM" src="https://github.com/user-attachments/assets/3f129422-e52c-4a35-8e7d-b980f89f0898">

But everything has become slow, since the total number of slots and the used slots have gone up significantly. Which is why AC has had to increase concurrency. The slots are used *after* waiting on the per-store AC queues (for regular work) and for eval flow tokens (for elastic work), so this slowness is not due to queueing due to AC or RAC. 
<img width="849" alt="Screenshot 2024-11-08 at 2 31 46 PM" src="https://github.com/user-attachments/assets/b552a9bb-1e08-47be-aa9a-1c1cad77af78">

Liveness heartbeat latency is also slow
<img width="849" alt="Screenshot 2024-11-08 at 2 45 01 PM" src="https://github.com/user-attachments/assets/fddd01ad-1b49-4c5e-ba95-ca4a5a6bf5af">

The p100 WAL fsync latency has spiked 
<img width="1311" alt="Screenshot 2024-11-08 at 2 29 52 PM" src="https://github.com/user-attachments/assets/d2d8600e-5d4d-4360-9bae-55fa6af25109">

But Pebble is healthy in terms of sub-levels etc.

The write bandwidth is declining but the write IOPS are increasing -- though this may not be relevant.
<img width="845" alt="Screenshot 2024-11-08 at 2 35 49 PM" src="https://github.com/user-attachments/assets/d9b60176-5e0c-4ffb-85e2-4cd511b25002">

**Comment by sumeerbhola** [comment]
I'm going to remove the AC label, since it doesn't seem to be an AC problem.

@andrewbaptist can you confirm that there is ample headroom wrt the provisioned disk bandwidth and IOPS. If not, it is possible that we got more than we deserved for some period of time, and then started getting throttled, which could explain these symptoms. You could also try with setting the provisioned bandwidth value, so AC knows that it should also shape elastic work to not saturate disk bandwidth.

**Comment by andrewbaptist** [comment]
This test is using local SSDs so it shouldn't be subject to cloud disk throttling. The drives have a relatively high peak bandwidth, although this variation was using larger values (4KB).
